File: leetcode-everyday/双指针/76-最小覆盖子串/76-最小覆盖子串.js
/**
 * @param {string} s
 * @param {string} t
 * @return {string}
 */
 var minWindow = function(s, t) {
  let chars = new Array(128).fill(0);
  let flag = new Array(128).fill(false);
  for(let i = 0; i < t.length; i++) {
      chars[t.charCodeAt(i)] ++;
      flag[t.charCodeAt(i)] = true;
  }
  
  let left = 0;
  let right = 0;
  let valid = 0;
  let start = 0;
  let end = 0;
  let len = Infinity;
  
  while(right < s.length) {
      let code = s.charCodeAt(right);
      if(flag[code]) {
          if(-- chars[code] >= 0) {
              ++ valid;
          }                        
          while(valid === t.length) {            
              if((right - left + 1) < len) {                    
                  len = right - left + 1;
                  start = left;
                  end = right;
              }
              if(flag[s.charCodeAt(left)] && ++chars[s.charCodeAt(left)] > 0) {
                  -- valid;
              }
              ++left;
          }
      }
      
      right ++;
  }
  return len === Infinity ? '' : s.substring(start, end+1);
};

File: leetcode-everyday/双指针/76-最小覆盖子串/76-最小覆盖子串.cpp
class Solution {
public:
    string minWindow(string s, string t) {
        vector<int> chars(128, 0);
        vector<bool> flag(128, false);
        for(int i = 0; i < t.size(); ++i) {
            ++chars[t[i]];
            flag[t[i]] = true;
        }
        int left = 0, right = 0, start = 0, end = 0, valid = 0, len = INT_MAX;
        while(right < s.size()) {
            if(flag[s[right]]) {
                if(--chars[s[right]] >= 0) {
                    ++valid;
                }
                while(valid == t.size()) {
                    if((right - left + 1) < len) {
                        len = right - left + 1;
                        start = left;
                        end = right;
                    }
                    if(flag[s[left]] && ++chars[s[left]] > 0) {
                        --valid;
                    }
                    ++left;
                }
            }
            right ++;
        }        
        return len == INT_MAX ? "" : s.substr(start, len);
    }
};